test/performance/odp_ml_perf.c

```diff
@@ -0,0 +1,1051 @@
+			ret = -1;
+			goto error;
+		}
+
+		output_final = output_file;
+		out_size_final = glb->out_size_d;
+	}
+
+	for (int i = 0; i < glb->num_out; i++) {
+		out_seg[i].addr = out_addr;
+		out_seg[i].size = glb->out[i].size;
+		out_addr += glb->out[i].size;
+	}
+
+	odp_ml_data_t data = {
+		.input_seg = inp_seg,
+		.num_input_seg = glb->num_inp,
+		.output_seg = out_seg,
+		.num_output_seg = glb->num_out,
+	};
+	odp_ml_run_param_t run_param;
+
+	odp_ml_run_param_init(&run_param);
+	run_param.batch_size = glb->opt.num_batch;
+
+	if (glb->opt.mode != MODE_INFERENCE_QUANT && glb->opt.scale_q > 0.0) {
+		if (quantize_input(input, glb->inp_file_data)) {
+			ret = -1;
+			goto error;
+		}
+	}
+
+	odp_barrier_wait(&glb->barrier);
+
+	for (int i = 0; i < glb->opt.rounds + glb->opt.warmup; i++) {
+		if (i == glb->opt.warmup)
+			time_start(&time);
+
+		if (glb->opt.mode == MODE_INFERENCE_QUANT && glb->opt.scale_q > 0.0) {
+			if (quantize_input(input, glb->inp_file_data)) {
+				ret = -1;
+				goto error;
+			}
+		}
+
+		int r;
+
+		while (!(r = odp_ml_run(glb->mdl, &data, &run_param)))
+			;
+		if (r != 1) {
+			ODPH_ERR("odp_ml_run() failed\n");
+			ret = -1;
+			goto error;
+		}
+
+		if (glb->opt.mode == MODE_INFERENCE_QUANT && glb->opt.scale_d > 0.0) {
+			if (dequantize_output(output_file, output)) {
+				ret = -1;
+				goto error;
+			}
+		}
+
+		if (glb->opt.latency && i >= glb->opt.warmup)
+			time_elapsed(&time, thread_idx);
+	}
+
+	if (!glb->opt.latency)
+		time_elapsed(&time, thread_idx);
+
+	if (glb->opt.mode != MODE_INFERENCE_QUANT && glb->opt.scale_d > 0.0) {
+		if (dequantize_output(output_file, output)) {
+			ret = -1;
+			goto error;
+		}
+	}
+
+	if (glb->ref_file_data) {
+		if (out_size_final != glb->ref_file_size) {
+			ODPH_ERR("Output size mismatch: %" PRIu64
+				 " differs from reference file size %" PRIu64 "\n",
+				 out_size_final, glb->ref_file_size);
+			ret = -1;
+			goto error;
+		}
+
+		if (memcmp(glb->ref_file_data, output_final, out_size_final)) {
+			ODPH_ERR("Output differs from reference\n");
+			ret = -1;
+			goto error;
+		}
+	}
+
+error:
+	free(input);
+	free(output);
+	free(output_file);
+
+	return ret;
+}
+
+static int test_ml_create(void *ptr)
+{
+	odp_time_t time;
+	int thread_idx = (int)(uintptr_t)ptr;
+	odp_ml_model_t mdl;
+	odp_ml_model_param_t model_param;
+	int ret = 0;
+
+	odp_ml_model_param_init(&model_param);
+	model_param.model = glb->model_file_data;
+	model_param.size = glb->model_file_size;
+
+	odp_barrier_wait(&glb->barrier);
+
+	for (int i = 0; i < glb->opt.rounds + glb->opt.warmup; i++) {
+		if (i == glb->opt.warmup)
+			time_start(&time);
+
+		mdl = odp_ml_model_create(glb->opt.model_name, &model_param);
+		if (mdl == ODP_ML_MODEL_INVALID) {
+			ODPH_ERR("odp_ml_model_create() failed\n");
+			ret = -1;
+			goto error;
+		}
+
+		if (odp_ml_model_destroy(mdl)) {
+			ODPH_ERR("odp_ml_model_destroy() failed\n");
+			ret = -1;
+			goto error;
+		}
```

Comment:
For stuff that's being tested, it's probably better to record their failures as some error counter increases instead of failing the thread. Same thing for loading and unloading in `test_ml_load()`.

Reply:
Again, better to just quit. We're measuring speed, not e.g. the probability of success.

Reply:
Well, in a performance testing setting, speed and probability of success might be related. Anyways, up to you.